Scale cosine distance to 0-1

hanit-com opened the pull request:
Fix - "_similarity_search_with_relevance_scores" is called by VectorStore base class and should return relevance score between 0 and 1, not the distance.

<!--
Thank you for contributing to LangChain! Your PR will appear in our release under the title you set. Please make sure it highlights your valuable contribution.

Replace this with a description of the change, the issue it fixes (if applicable), and relevant context. List any dependencies required for this change.

After you're done, someone will review your PR. They may suggest improvements. If no one reviews your PR within a few days, feel free to @-mention the same people again, as notifications can get lost.

Finally, we'd love to show appreciation for your contribution - if you'd like us to shout you out on Twitter, please also include your handle!
-->

<!-- Remove if not applicable -->

Fixes #6481 

#### Before submitting

<!-- If you're adding a new integration, please include:

1. a test for the integration - favor unit tests that does not rely on network access.
2. an example notebook showing its use


See contribution guidelines for more information on how to write tests, lint
etc:

https://github.com/hwchase17/langchain/blob/master/.github/CONTRIBUTING.md
-->

#### Who can review?

Tag maintainers/contributors who might be interested: @dev2049 

<!-- For a quicker response, figure out the right person to tag with @

  @hwchase17 - project lead

  Tracing / Callbacks
  - @agola11

  Async
  - @agola11

  DataLoaders
  - @eyurtsev

  Models
  - @hwchase17
  - @agola11

  Agents / Tools / Toolkits
  - @hwchase17

  VectorStores / Retrievers / Memory
  - @dev2049

 -->

hanit-com committed:
Scale cosine distance to 0-1 (#6481)

Fix - "_similarity_search_with_relevance_scores" is called by VectorStore base class and should return relevance score between 0 and 1, not the distance.

dev2049 commented on langchain/vectorstores/chroma.py:
```diff
@@ -228,7 +228,23 @@ def _similarity_search_with_relevance_scores(
+        """
+
+        docs_and_scores = self.similarity_search_with_score(query, k)
+        return [(doc, 1.0 - score) for doc, score in docs_and_scores]
```
i think chroma scores can also be squared l2 distance, which isn't normalized. we should figure out a way to get the distance for the collection and normalize based on that

hanit-com committed:
Support different normalization methods (#6481).

- Normalization methods for "l2", "ip", and "cosine".
- Setting different distance methods when creating Chroma vectorstore from texts and documents.
- Support in custom normalization method provided by the user.

hanit-com committed:
Removed redundancy.

Pass "**kwargs" instead of relevance_score_fn arameter.

raymond-yuan commented:
Can we add this to the base VectorStore class or add a specialized subclass? It seems FAISS already has this implemented: https://github.com/hwchase17/langchain/blob/master/langchain/vectorstores/faiss.py#L81 with slightly different relevant scores though.

raymond-yuan commented:
This includes some of the changes you've included in this PR, do you want to take a look? Happy to collaborate/merge your additional changes in if we think it's worth refactoring in this way:

https://github.com/hwchase17/langchain/pull/6570

hanit-com commented:
Hi @raymond-yuan, the metric scores may differ between the databases, for example, pgvector returns the negative of maximum inner product while chroma will return a positive value for this metric, so the normalization/ scaling function should match the database.
I think the default **relevance_score_fn** should be left unimplemented in base for the specific vector store to implement. Besides that I think it is a necessary refactor and a great job in #6570.
I have two pull requests that include related code - this one and #6475. Those versions are working well and the normalization functions I added for chroma and pgvector are tested.
You want to take my chroma and pgvector normalization functions to your pull request and I will resolve the conflicts and match my PRs to the new code? Or the other way around when I merge my changes and the refactor will be done after?
Let me know.

raymond-yuan commented:
>  default relevance_score_fn should be left unimplemented in base
I think this should be fine right? We let the vector store define whatever relevance score in `_select_relevance_score_fn` and we rely on that function in `_similarity_search_with_relevance_scores`
https://github.com/hwchase17/langchain/blob/fe0739b42e8030544d6c95618896fc6f118d9399/langchain/vectorstores/base.py#L184
or maybe I'm misunderstanding your point?

> You want to take my chroma and pgvector normalization functions to your pull request and I will resolve the conflicts and match my PRs to the new code? Or the other way around when I merge my changes and the refactor will be done after? Let me know.

Either way works! I think maybe the easiest approach would be to merge your changes in to my PR since it is a bit of a superset of changes? Let me know if you want me to go ahead and merge the two PRs into mine

hanit-com commented:
I like this suggestion, nevertheless, each vector DB returns a different formula, there might be vectorstore that will return the inner product already normalized or return (1.0 - max_inner product). It is a broad subject to resolve in one pull request so I think it is better to handle it in vector store level while keeping each vector store's normalization function as it was.
The centralized normalization functions solution can work if you are confident about it, and it might require adjusting the tests to match the new scores (if changed).
It is also a good thing the user can provide his own implementation in case we got it inaccurately.
About the pull requests, if it works for you then sure, you can merge them into yours!

raymond-yuan commented:
I think it's good to provide some default functions because most of the implementations are standardized. The current implementation does not prevent users from specifying different relevance functions for specific vectorbases, for example one could directly pass the score through the _select_relevance_score_fn without normalizing it if the db has already done this. As of now all of the normalization functions work with the tests as is. We could potentially move the centralized functions into vectorstores/utils, but in either case because so many implementations rely on relatively standard functions, i think providing some default functions for the relevance functions helps a lot with code duplication. 

I went ahead and merged your changes in to my PR (and handled conflicts)!

hanit-com commented:
Merged into #6570.
